libobs-opengl/gl-egl-common.c

```diff
@@ -215,6 +217,125 @@ gl_egl_create_dmabuf_image(EGLDisplay egl_display, unsigned int width,
+		blog(LOG_ERROR, "Unable to load eglQueryDmaBufFormatsEXT");
+		return 0;
+	}
+
+	if (!glad_eglQueryDmaBufFormatsEXT(egl_display, 0, NULL,
+					   &max_formats)) {
+		blog(LOG_ERROR, "Cannot query the number of formats: %s",
+		     gl_egl_error_to_string(eglGetError()));
+		return 0;
+	}
+
+	format_list = bzalloc(max_formats * sizeof(EGLint));
+	if (!format_list) {
+		blog(LOG_ERROR, "Unable to allocate memory");
+		return 0;
+	}
+
+	if (!glad_eglQueryDmaBufFormatsEXT(egl_display, max_formats,
+					   format_list, &max_formats)) {
+		blog(LOG_ERROR, "Cannot query a list of formats: %s",
+		     gl_egl_error_to_string(eglGetError()));
+		free(format_list);
+		return 0;
+	}
+
+	*formats = format_list;
+	return max_formats;
+}
+
+EGLint query_dmabuf_modifiers(EGLDisplay egl_display, EGLint drm_format,
+			      EGLuint64KHR **modifiers)
+{
+	EGLint max_modifiers = 0;
+	EGLuint64KHR *modifier_list = NULL;
+	EGLBoolean *external_only = NULL;
+
+	if (!glad_eglQueryDmaBufModifiersEXT) {
+		blog(LOG_ERROR, "Unable to load eglQueryDmaBufModifiersEXT");
+		return 0;
+	}
+
+	if (!glad_eglQueryDmaBufModifiersEXT(egl_display, drm_format, 0, NULL,
+					     NULL, &max_modifiers)) {
+		blog(LOG_ERROR, "Cannot query the number of modifiers: %s",
+		     gl_egl_error_to_string(eglGetError()));
+		return 0;
+	}
+
+	modifier_list = bzalloc(max_modifiers * sizeof(EGLuint64KHR));
+	external_only = bzalloc(max_modifiers * sizeof(EGLBoolean));
+	if (!modifier_list || !external_only) {
+		blog(LOG_ERROR, "Unable to allocate memory");
+		bfree(external_only);
+		bfree(modifier_list);
+		return 0;
+	}
+
+	if (!glad_eglQueryDmaBufModifiersEXT(egl_display, drm_format,
+					     max_modifiers, modifier_list,
+					     external_only, &max_modifiers)) {
+		blog(LOG_ERROR, "Cannot query a list of modifiers: %s",
+		     gl_egl_error_to_string(eglGetError()));
+		bfree(external_only);
+		bfree(modifier_list);
+		return 0;
+	}
+
+	bfree(external_only);
+	*modifiers = modifier_list;
+	return max_modifiers;
+}
+
+int gl_egl_query_dmabuf_modifiers(EGLDisplay egl_display, uint32_t drm_format,
+				  uint64_t **modifiers)
+{
+	if (!glad_eglQueryDmaBufFormatsEXT || !glad_eglQueryDmaBufModifiersEXT)
+		return 0;
+
+	EGLint num_formats;
+	EGLint *format_list;
+	bool format_available;
+
+	if (!(num_formats = query_dmabuf_formats(egl_display, &format_list))) {
+		blog(LOG_ERROR, "No formats supported by dmabuf");
+		return 0;
+	}
+
+	for (int i = 0; i < num_formats; i++) {
+		if (drm_format == (uint32_t)format_list[i]) {
+			format_available = true;
+			break;
+		}
+	}
+	bfree(format_list);
+
+	if (!format_available) {
+		blog(LOG_ERROR, "Format %u not supported with modifiers",
+		     drm_format);
+		return 0;
+	}
+
+	EGLint num_modifiers;
+	EGLuint64KHR *modifier_list;
+
+	num_modifiers =
+		query_dmabuf_modifiers(egl_display, drm_format, &modifier_list);
+
+	brealloc(modifier_list, (num_modifiers + 1) * sizeof(uint64_t));
+	modifier_list[num_modifiers] = DRM_FORMAT_MOD_INVALID;
```

Comment:
This feels somewhat like a thing that should be handled in another layer, e.g. in `build_format()` in `linux-capture: Announce supported modifiers via PipeWire`.

Reply:
It can't be done in `build_format()`. The plugin doesn't have any clue which renderer is used. If obs would add a vulkan renderer in the future adding the token for implicit modifier would be false. So we either add it at the end of the list returned by query_modifiers or we need another call to the renderer to query if implicit modifiers are supported.

Reply:
My personal choice would be to let the renderer tell whether implicit modifiers are supported, so that the PipeWire bits can deal with this particularity.

Reply:
I tried minimizing the change in the renderer API, but I'm totally fine with adding a separate call returning a bool for implicit modifier supported, keeping the query function modifier aware only.

Reply:
Maybe best consult maintainers, I'm just a fly-by reviewer here :P

Reply:
Sure, my comment is an offer to the maintainers. Just wanted to acknowledge, that I had the same thoughts and decided on one function with both explicit and implicit combined.

Reply:
Looking at the previous commits, it seems that modifierless is already handled in the pipewire code separately as suggested. Is the point here to try and merge the modifierless and modifierfull code paths? If so we should remove the modifierless checks in the pipewire code.

Reply:
Dma-Bufs and modifiers are a rather nasty and complicated topic. I'll try to summarize it bellow in a comment.

Reply:
After https://github.com/obsproject/obs-studio/pull/5131#issuecomment-980476360

The question here is: 
Do we want to reflect the different kind of API's with the query_modifiers function only representing the modifier-aware API call and having a seperate call to check if implicit modifiers are available, or do we want to merge those two into one call applying the placeholder as described below.

Reply:
Generally I would prefer we export with the same semantics as the egl/gl functions we are calling unless we have a good reason to modify them for keeping state consistent in libobs. For these platform specific bits I think theres even more reason to not introduce our own semantics.

While it seems convenient to build the whole list in of formats and their modifiers but its probably safest to simply export the QueryFormat and QueryModifier functions separately on the graphics subsystem and build these lists in the pipewire code.

Reply:
Ok so I'll design the API to be close to the egl functions and add additional bool functions to check for available extensions.

Reply:
Do we need to provide a way to  query extensions? If we return no modifiers when obs detects the extensions are not available then users can fallback to modifierless gracefully without the additional check? Or is there still a need to know if the driver supports the modifier extensions?

Reply:
Support for implicit modifiers in EGL is defined by the existence of an extension. But I think a bool function `_implicit_modifier_available` would be better, checking that required extension internally.

Reply:
Is the new API with 3 functions close enough to the EGL functions or do you want the same signature?